src/main/geometries/GeometryList.java

```diff
@@ -68,24 +78,47 @@ public void add(GeometryList geometries) {
-		// TODO: implement
 		// move all finite geometries to a sub-geometries (use g.border().isFinite() to determine which ones to add)
+		GeometryList finiteGeometries = new GeometryList();
+		intersectibles.removeIf(i -> {
+			if(i.border().isFinite()) {
+				finiteGeometries.add(i);
+				return true;
+			}
+			return false;
+
+		});
 		finiteGeometries.constructHierarchy();
-                this.intersectibles.add(finiteGeometries);
+		this.intersectibles.add(finiteGeometries);
 	}
 
 	private void constructHierarchy() {
 		// TODO: implement
 		// this is where the fun begins
-                // An overview of an algorithm might go as follows
-                // 1. For each unordered pair of geometries {a,b} calculate the SA of the Union of their bounding boxes
-                // 2. Store these values in some data structure (we'd have to find/make one to efficiently support our operations, possibly some sort of graph)
-                // 3.1. Remove the pair {a,b} with the smallest bounding box.
-                // 3.2. Remove all pairs {x,a} or {x,b} (i.e. all paid which contain a or b)
-                // 4. Make a GeometryList c which contains a and b (note, we can pass the already computed bounding box of a and b into a private constructor so it won't need to recompute it). 
-                // 5. For each remaining Intersectible x add the pair {c,x} to the data structure and compute the bounding box of c union X.
-                // 6. Repeat until there are two Intersectibles i1 and 12 left.
-                // 7. this.intersectibles.add(i1, i2)
+		// An overview of an algorithm might go as follows
+		// 1. For each unordered pair of geometries {a,b} calculate the SA of the Union of their bounding boxes
+		// 2. Store these values in some data structure (we'd have to find/make one to efficiently support our
+		// operations, possibly some sort of graph)
+		// 3.1. Remove the pair {a,b} with the smallest bounding box.
+		// 3.2. Remove all pairs {x,a} or {x,b} (i.e. all paid which contain a or b)
+		// 4. Make a GeometryList c which contains a and b (note, we can pass the already computed bounding box of a and
+		// b into a private constructor so it won't need to recompute it).
+		// 5. For each remaining Intersectible x add the pair {c,x} to the data structure and compute the bounding box
+		// of c union X.
+		// 6. Repeat from step 3 until there are two Intersectibles i1 and 12 left.
+		// 7. this.intersectibles.add(i1, i2)
+		//
+		WeightedGraph<Intersectible, BoundingBox> G =
+			new WeightedGraph<>(intersectibles, (i1, i2) -> i1.border().union(i2.border()));
+		WeightedGraph<Intersectible, BoundingBox>.Edge minEdge; // a typedef would be nice here
+		while(G.size() >= 2) {
+			minEdge = G.extract();
+			G.add(new GeometryList(minEdge.weight, minEdge.vertex1, minEdge.vertex2));
+		}
+		minEdge = G.extract();
+		this.intersectibles.add(minEdge.vertex1);
+		this.intersectibles.add(minEdge.vertex2);
+		this.border = minEdge.weight;
```

Comment:
You don't need this step. This border is already as tight as it can be. Rearranging its geometries won't change its border